src/thread.ts

```diff
@@ -140,226 +146,200 @@ export class Thread<SCG extends ExtendableGenerics = DefaultGenerics> {
-        console.warn(error);
-      } finally {
-        // this.updateLocalState('recovering', false);
-      }
-    };
-
-    // when the thread becomes active or it becomes stale while active (channel stops being watched or connection drops)
-    // the recovery handler pulls its latest state to replace with the current one
-    // failed messages are preserved and appended to the newly recovered replies
-    this.unsubscribeFunctions.add(
-      this.state.subscribeWithSelector(
-        (nextValue) => [nextValue.active, nextValue.isStateStale],
-        ([active, isStateStale]) => {
-          if (active && isStateStale) handleStateRecovery();
-        },
-      ),
+  private subscribeReloadActiveStaleThread = () =>
+    this.state.subscribeWithSelector(
+      (nextValue) => [nextValue.active, nextValue.isStateStale],
+      ([active, isStateStale]) => {
+        if (active && isStateStale) {
+          this.loadState();
+        }
+      },
     );
 
-    this.unsubscribeFunctions.add(
-      // TODO: figure out why the current user is not receiving this event
-      this.client.on('user.watching.stop', (event) => {
-        if (!event.channel || !event.user || !this.client.userID || this.client.userID !== event.user.id) return;
-
-        const { channel } = this.state.getLatestValue();
-
-        if (event.channel.cid !== channel.cid) return;
-
-        this.state.partialNext({ isStateStale: true });
-      }).unsubscribe,
-    );
+  private subscribeMarkThreadStale = () =>
+    this.client.on('user.watching.stop', (event) => {
+      const { channel } = this.state.getLatestValue();
 
-    this.unsubscribeFunctions.add(
-      this.client.on('message.new', (event) => {
-        if (!event.message || !this.client.userID) return;
-        if (event.message.parent_id !== this.id) return;
+      if (!this.client.userID || this.client.userID !== event.user?.id || event.channel?.cid !== channel.cid) {
+        return;
+      }
 
-        const { isStateStale } = this.state.getLatestValue();
+      this.state.partialNext({ isStateStale: true });
+    }).unsubscribe;
 
-        if (isStateStale) return;
+  private subscribeNewReplies = () =>
+    this.client.on('message.new', (event) => {
+      if (!this.client.userID || event.message?.parent_id !== this.id) {
+        return;
+      }
 
-        if (this.pendingRepliesMap.has(event.message.id)) {
-          this.pendingRepliesMap.delete(event.message.id);
+      const isOwnMessage = event.message.user?.id === this.client.userID;
+
+      this.upsertReplyLocally({
+        message: event.message,
+        // Message from current user could have been added optimistically,
+        // so the actual timestamp might differ in the event
+        timestampChanged: isOwnMessage,
+      });
+
+      const { read } = this.state.getLatestValue();
+      const readUpdate: ThreadReadState = {};
+
+      for (const userId of Object.keys(read)) {
+        if (read[userId]) {
```

Comment:
Not sure I understand this part - if we can target specific `userId` (based on `event.user`) in a pre-formatted read map then why going over all of the other users?

Reply:
That's because a new message should be considered unread by everybody (we know of) until the `message.read` event arrives.

`event.user` will be the author of the message, and we have to increment the unread count for everybody else.

See how Channel handles `message.new`.